### ISSUE TITLE
On a mix the melody is lost to separation, and what the loss is made of is not yet known

### ISSUE THREAD

**Issue by eolivelli** [eolivelli/music-wizard#503]
Measured by `tools/measure-separation-cost.py`, which scores annotated solo
voices three ways against the same annotations: clean, through `mw separate`
with no band at all, and through `mw separate` after a band has been mixed in.

Ten `vocadito` clips, the committed synthetic bed:

| | note F1 @50 ms | raw pitch accuracy | voicing recall |
|---|---|---|---|
| clean voice | 65.8% | 84.0% | 97.0% |
| separator only, no band | 64.5% | 83.9% | 97.2% |
| band mixed in, then separated | 23.2% | 37.2% | 79.3% |

**The separator does not spoil a voice by itself.** The middle row costs about
a point of note F1 and nothing measurable in pitch.

**The whole loss appears once a band is present, and this tool cannot say what
it is made of.** The middle row is a null control — with no band there is
nothing for the mask to remove — so it bounds what the separator does to a
voice alone and apportions nothing. The gap between the outer rows holds two
different defects mixed together:

- band the mask failed to remove, which leaves the tracker something else to
  follow, and
- voice the mask removed along with the band, which leaves it nothing.

They want opposite fixes, and an earlier version of this issue asserted the
first and used it to rule out remedies for the second. Projecting the separated
stems back onto the clean voice says the second is large at the ratios these
clips present: on the clips that lose the most, the voice comes out of the
separator 20 to 30 dB down. So nothing here rules out a different checkpoint —
4- or 5-stem, or higher bandwidth — and the earlier text saying it did was
wrong.

**How large any of this is depends on the vocal-to-band ratio, which the tool
does not control (#505).** The clips whose voice sits well under the bed are
exactly the ones that lose everything, and which clips those are is an accident
of how loudly each singer was recorded. Read the third row as "what this path
costs at whatever ratios these ten clips happen to present", and fix #505
before comparing beds or checkpoints against each other.

What survives all of that: on a mix, the melody stage's accuracy is a statement
about separation rather than about `PitchTracker` or `MelodyEstimator`. On a
clean voice the stage sits close under the annotators' own agreement (see
`tools/baselines/score-melody-vocadito.txt`), which is what closed #502.

**Comment by eolivelli** [comment]
#559 changed what feeds this question, and added a measurement it can use.

`analyze --melody` now reads the separated vocal wherever a separation
provider can be had, so the gap this issue is about is no longer hypothetical
for the melody stage — it is in the product path, and `--skip-separation` is
the way to take it out again.

Two committed baselines make the separator's cost to melody directly
measurable, on the same corpora read both ways
(`tools/score-melody.py --separated`):

- `score-melody-vocadito.txt` vs `score-melody-vocadito-separated.txt` — 40
  clips of real solo singing, separated and not. This is the clean measurement
  of "what does the separator do to a voice that is alone": individual clips
  move both ways and no column's mean moves by as much as a point, which is
  this issue's own finding, now reproducible by a harness rather than by a
  one-off.
- `score-melody.txt` vs `score-melody-separated.txt` — the synthetic packages.
  Two things there: the packages with a band under the melody, where
  separating is the only reason the pitch column is above the floor at all,
  and the melody-only packages, where the melody is a *rendered instrument*
  playing alone and largely does not survive a vocal separator. That second
  group is a cost this issue may want to own: it is separation quality on
  non-vocal material, not the melody stage.

Both are local-only (they need the Spleeter model in the cache) and diffed by
`tools/premerge.sh` steps 9 and 10; a machine without the model skips every
row rather than failing.

Choosing between the mix and the stem by evidence instead of by a flag is
#560, and its answer depends on what this issue concludes.

**Comment by eolivelli** [comment]
**FIX_AS_DESCRIBED.** The projection this issue named is now a re-runnable
harness, and it answers the question: **the loss is voice the mask removed,
not band it left behind.** PR #573.

### How it is apportioned

The measurement makes the mix, so the band inside it is `mix - voice` sample
for sample, where the separator only saw their sum.
`tools/SeparationSplit.java` divides every bin of the stem between the two
sources in proportion to the energy each has there — so the parts add back to
the stem, and the run asserts that rather than assuming it — and
`tools/apportion-separation-loss.py` scores each part against the same
annotations.

Five vocadito clips, the committed bed, note F1 at 50 ms. The other five rail
before the bed is loud (#518):

| voice over bed | clean | mix | voice part | + band part | stem |
|---|---|---|---|---|---|
| −12 dB | 75.0 | 0.0 | 20.5 | 61.4 | 10.6 |
| −6 dB | 75.0 | 0.0 | 38.4 | 64.7 | 30.8 |
| 0 dB | 75.0 | 4.9 | 55.2 | 69.6 | 47.8 |
| +6 dB | 71.2 | 32.1 | 62.2 | 66.6 | 56.5 |

`voice part` is what the separator delivered of the voice, alone; `+ band
part` is the untouched voice with the stem's whole leftover band added back.
Read them as two subtractions from `clean`, not as a chain. The `mix` row is
the floor: separation is worth running by a wide margin, whatever it costs.

The energy figures the split prints do **not** predict this. Both checkpoints
below lose the voice about equally in decibels and differ by a lot in what the
tracker can then do with it, which is why the parts are scored rather than
merely measured.

### What the answer bought

Spleeter squares the estimates before taking their ratio, which cuts hardest
exactly where the band beats the voice in a bin — paying for rejection with
retention, the wrong way round for this stage. #573 takes the ratio of the
estimates themselves. Stem row, same protocol, note F1 / raw pitch:

| voice over bed | squared | plain |
|---|---|---|
| −12 dB | 10.4 / 12.7 | 10.6 / 15.1 |
| −6 dB | 24.0 / 34.3 | 30.8 / 39.5 |
| 0 dB | 45.0 / 67.4 | 47.8 / 70.3 |
| +6 dB | 55.6 / 79.6 | 56.5 / 80.6 |

`tools/baselines/score-melody-separated.txt` moved with it, every band-backed
and every melody-only package improving on the pitch column;
`score-melody-vocadito-separated.txt` moved both ways by under a point, which
is what a corpus with no band under it should do.

### The 4-stem checkpoint: measured, declined

This issue left 4- and 5-stem checkpoints open, so:

**Spleeter 4stems is licence-clean and does exist as ONNX.** Deezer's weights
under the same JOSS statement as the 2stems we ship (doi:10.21105/joss.02154,
"source code and pre-trained models ... under a MIT license"), exported
through sherpa-onnx's Apache-2.0 U-Net by
`Best-Practice/spleeter-4stems-onnx` on Hugging Face, whose conversion scripts
are published. Same input signature as our 2stems export, so it runs on this
provider with only the mask generalised over more stems. sha256, in case
anyone repeats this:

```
vocals 341bd47da8a0ea590178e9d342cbd6ac168cf05f0bf8ce6a62d1ef18f8358701
drums  fe832ff1b9ea54b7d3fd8256fa464e21dbccb3cafa83d4f72d3ee2d5dc21f997
bass   be84af108ec6a5b538d35fd5643799393c4c0bbfa21f2ea1308ca1364a076d81
other  9fa7fc4ecbe8d75f0b3c9fea65a0c8b7a32518a7c0ed55339036ad342fb8bedd
```

**It loses.** Measured on the protocol above and on all 40 vocadito clips,
from a local patch that is deliberately not in #573 — so unlike the tables
above, these do not reproduce from the repository:

| voice over bed | 2stems | 4stems |
|---|---|---|
| −12 dB | 10.4 / 12.7 | 8.0 / 10.8 |
| −6 dB | 24.0 / 34.3 | 23.5 / 28.8 |
| 0 dB | 45.0 / 67.4 | 40.9 / 54.2 |
| +6 dB | 55.6 / 79.6 | 56.6 / 75.7 |

On solo voice through the separator the two are indistinguishable (mean note
F1 61.5% against 61.9%, pitch 82.1% against 82.1%). The split says why: 4stems
rejects the band harder and cuts more voice, which is the wrong side of the
trade this issue's measurement identifies. **5stems has no published ONNX
export at all**, and by the same mechanism should be worse still — more
competing stem models, a more selective vocal mask. Bandwidth is not on offer
either: both published exports declare the same 1024-bin grid, and a sung
fundamental sits far below that ceiling.

### Left open

- #575 — the voice is still deleted where the band is loud, and nothing
  licence-clean we have measured does better.
- #518 and #515 — the corpus cannot reach the loud-band regime, and absolute
  level is a second uncontrolled axis. Both bound everything above.
- #574 — only the two ends of the mask exponent were tried.

One more thing found by running it: `tools/measure-separation-cost.py` had not
run at all since #559 changed what `analyze` returns. Fixed in #573, with a
rule test, since nothing else calls that function without a jar.
